collector/vmstat_numa_collector.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build !nonumavmstat
+
+
+package collector
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/go-kit/kit/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+)
+
+const vmStatNumaSubsystem = "vmstat_numa"
+
+type vmstatNumaCollector struct {
+	metricDescs map[string]*prometheus.Desc
+	logger      log.Logger
+	fs          sysfs.FS
+}
+
+func init() {
+	registerCollector("vmstat_numa", defaultDisabled, NewVmstatNumaCollector)
+}
+
+// NewVmstatNumaCollector returns a new Collector exposing memory stats.
+// Returns an error when cAdvisor can't read procfs.
+func NewVmstatNumaCollector(logger log.Logger) (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+	return &vmstatNumaCollector{
+		metricDescs: createMetricDescriptions(),
+		logger:      logger,
+		fs:          fs,
+	}, nil
+}
+
+func (c *vmstatNumaCollector) Update(ch chan<- prometheus.Metric) error {
+
+	metrics, err := c.fs.VMStatNUMA()
+	if err != nil {
+		return fmt.Errorf("couldn't get NUMA vmstat: %w", err)
+	}
+	for k, v := range metrics {
+		metricStruct := reflect.ValueOf(v)
+		typeOfMetricStruct := metricStruct.Type()
+		for i := 0; i < metricStruct.NumField(); i++ {
+			desc := c.metricDescs[typeOfMetricStruct.Field(i).Name]
+			ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, float64(metricStruct.Field(i).Uint()),
+				strconv.Itoa(k))
+		}
+
+	}
+	return nil
+}
+
+func createMetricDescriptions() map[string]*prometheus.Desc {
+	return map[string]*prometheus.Desc{
```

Comment:
In general, maybe we can merge some of these metrics in a timeseries with labels?
Like, can the pages metrics be summed up to give a sensible 'number of available pages' that can be split down to have `nr_pages{state="free"} ` and `nr_pages{state="active_file"}` and so on?